### PR TITLE
Fix lodash imports

### PR DESCRIPTION
**Pull request by resynth1943**
Changed some `lodash` imports to `lodash/...`. Hope this pushes the bundle size down a little, since it seems like the entirety of Lodash was being imported.

### CODE REVIEW

**Comment by turt2live** [comment]
@resynth1943 can we get sign-off per the contributing guidelines? Thanks!

**Comment by resynth1943** [comment]
FFS. I legit went out of my way to sign it, looks like my .patch silliness wiped the sign-off 😓

**Comment by resynth1943** [comment]
# Before



## After



Looks like this PR knocks 5MB off the bundle size.

**Comment by resynth1943** [comment]
@turt2live go for it, just force-pushed again. Sorry for that 😅

File: src/components/structures/RoomSearch.tsx [hunk 1]
```diff
@@ -20,7 +20,6 @@ import classNames from "classnames";
 import defaultDispatcher from "../../dispatcher/dispatcher";
 import { _t } from "../../languageHandler";
 import { ActionPayload } from "../../dispatcher/payloads";
-import { throttle } from 'lodash';
```
Comment (resynth1943) [inline src/components/structures/RoomSearch.tsx:23]: Removed this because it wasn't used.

**Review by dbkr — COMMENTED** [review]
Are you sure you weren't testing a debug build against a production build? Seems very unlikely that tweaking lodash imports would reduce the bundle size by 5MB: I don't think the whole of lodash is anywhere near that large.

For me it doesn't make any difference. Before patch:
```
-rw-r--r--  1 dave  staff    13K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/0.js
-rw-r--r--  1 dave  staff   9.6K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/1.js
-rw-r--r--  1 dave  staff    19K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/2.js
-rw-r--r--  1 dave  staff   800K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/20.js
-rw-r--r--  1 dave  staff   7.0K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/21.js
-rw-r--r--  1 dave  staff   1.2K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/22.js
-rw-r--r--  1 dave  staff   3.2K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/23.js
-rw-r--r--  1 dave  staff   1.3K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/24.js
-rw-r--r--  1 dave  staff   2.0K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/25.js
-rw-r--r--  1 dave  staff   1.7K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/26.js
-rw-r--r--  1 dave  staff   2.1K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/27.js
-rw-r--r--  1 dave  staff    17K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/bundle.js
-rw-r--r--  1 dave  staff   2.9K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/compatibility-view.js
-rw-r--r--  1 dave  staff   1.0K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/error-view.js
-rw-r--r--  1 dave  staff    97K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/indexeddb-worker.js
-rw-r--r--  1 dave  staff   2.0M 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/init.js
-rw-r--r--  1 dave  staff    15K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/jitsi.js
-rw-r--r--  1 dave  staff   9.4K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/mobileguide.js
-rw-r--r--  1 dave  staff   3.6K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/riot-web-app.js
-rw-r--r--  1 dave  staff   2.9K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/riot-web-component-index.js
-rw-r--r--  1 dave  staff   982B 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/theme-dark-custom.js
-rw-r--r--  1 dave  staff   975B 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/theme-dark.js
-rw-r--r--  1 dave  staff   982B 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/theme-legacy-dark.js
-rw-r--r--  1 dave  staff   977B 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/theme-legacy.js
-rw-r--r--  1 dave  staff   983B 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/theme-light-custom.js
-rw-r--r--  1 dave  staff   976B 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/theme-light.js
-rw-r--r--  1 dave  staff   1.8K 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/usercontent.js
-rw-r--r--  1 dave  staff   2.8M 27 Aug 09:50 webapp/bundles/1386b15091a32760be55/vendors~init.js
```

with your patch:
```
-rw-r--r--  1 dave  staff    13K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/0.js
-rw-r--r--  1 dave  staff   9.6K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/1.js
-rw-r--r--  1 dave  staff    19K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/2.js
-rw-r--r--  1 dave  staff   800K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/20.js
-rw-r--r--  1 dave  staff   7.0K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/21.js
-rw-r--r--  1 dave  staff   1.2K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/22.js
-rw-r--r--  1 dave  staff   3.2K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/23.js
-rw-r--r--  1 dave  staff   1.3K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/24.js
-rw-r--r--  1 dave  staff   2.0K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/25.js
-rw-r--r--  1 dave  staff   1.7K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/26.js
-rw-r--r--  1 dave  staff   2.1K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/27.js
-rw-r--r--  1 dave  staff    17K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/bundle.js
-rw-r--r--  1 dave  staff   2.9K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/compatibility-view.js
-rw-r--r--  1 dave  staff   1.0K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/error-view.js
-rw-r--r--  1 dave  staff    97K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/indexeddb-worker.js
-rw-r--r--  1 dave  staff   2.0M 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/init.js
-rw-r--r--  1 dave  staff    15K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/jitsi.js
-rw-r--r--  1 dave  staff   9.4K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/mobileguide.js
-rw-r--r--  1 dave  staff   3.6K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/riot-web-app.js
-rw-r--r--  1 dave  staff   2.9K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/riot-web-component-index.js
-rw-r--r--  1 dave  staff   982B 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/theme-dark-custom.js
-rw-r--r--  1 dave  staff   975B 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/theme-dark.js
-rw-r--r--  1 dave  staff   982B 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/theme-legacy-dark.js
-rw-r--r--  1 dave  staff   977B 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/theme-legacy.js
-rw-r--r--  1 dave  staff   983B 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/theme-light-custom.js
-rw-r--r--  1 dave  staff   976B 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/theme-light.js
-rw-r--r--  1 dave  staff   1.8K 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/usercontent.js
-rw-r--r--  1 dave  staff   2.8M 27 Aug 09:54 webapp/bundles/83187f2d9fec80fd441e/vendors~init.js
```

**Comment by t3chguy** [comment]
image

^ is the lodash we include in the bundle, sure seems like a lot but I wonder how much of it is transitive deps/layered functions

image

But yeah there is no way it could have a 5M impact.

**Comment by resynth1943** [comment]
Hmm, curious...

**Comment by resynth1943** [comment]
Hmm, just reviewed locally, not sure where I got 5MB from 😅 That being said, it did remove 100KB, which is pretty cool.

**Comment by t3chguy** [comment]
@resynth1943 from @dbkr's test above it doesn't even achieve that though?

**Comment by resynth1943** [comment]
> @resynth1943 from @dbkr's test above it doesn't even achieve that though?

> Hmm, just reviewed locally

I literally made a fresh build from up-to-date sources. No idea what's happening above...

**Comment by t3chguy** [comment]
This breaks certain typescript definitions; seems like the `@types/lodash` package might prefer the other approach. 

Before:
image

After:
image

---

From testing it myself I actually find this to be larger than it was before.


Before:
image
image

After:
image
image

---

I thought to try changing them all to destructured imports, results below:

image

**Comment by resynth1943** [comment]
> From testing it myself I actually find this to be larger than it was before.

WTF are you testing tho? I did this locally and it shaved off 100KB. Literally pulled straight from the develop branch when I tested it. I honestly don't know what you guys are doing wrong...

I think we are testing different products, because the screenshots you listed don't represent the output of the compiled code from the develop branch.

I'll check the type thing tomorrow, but I doubt it's anhrjing serious if the Lodash types are worth their salt.

**Comment by turt2live** [comment]
As far as I can tell, this has no significant bearing on the bundle size: my before is in red here with the after in green: https://gist.github.com/turt2live/242babc4fa4322c03bb3039f75f5945a/revisions

For the benefit of doubt, I made sure to merge this PR locally instead of checking it out against whatever it's based on, as that can affect the bundle size.

Our version of webpack does automatic tree-shaking, and it does in fact seem to be doing it quite well.

**Comment by resynth1943** [comment]
Hmm. Have you tried this patch on Element Web? Like with this version as a dependency?

**Comment by turt2live** [comment]
That's not how development environments work. Because it's linked, verifiably, the code is being pulled in.

**Comment by turt2live** [comment]
Overall I don't think this actually does anything besides make types harder to use. No one on the team can reproduce significant savings, implying that webpack is doing its tree shaking extremely well.

Thank you for the contribution in any case - we'd be happy to accept PRs which reduce the bundle size, assuming they can do so safely.

**Comment by resynth1943** [comment]
> That's not how development environments work. Because it's linked, verifiably, the code is being pulled in.

Yes, I am aware of how Node's dependency resolution works 😉

**Comment by resynth1943** [comment]
> No one on the team can reproduce significant savings, implying that webpack is doing its tree shaking extremely well.

Alright. That's disappointing, since this did shave off 100KB in my testing. 

Have a splendid day!

**Pull request closed without merging**